vignettes/potential-optimizers.Rmd

```diff
@@ -0,0 +1,670 @@
+```
+
+#### Proposed Optimized Code
+```{r mat_df_code_op}
+
+mat_fun <- function(){
+  mat <- matrix(1:352, nrow = 32, ncol = 11)
+  mat[4, ]
+}
+```
+
+### Benchmark
+
+```{r mat_df_benchmark, warning = FALSE, message = FALSE}
+library("ggplot2")
+library("microbenchmark")
+autoplot(microbenchmark(df_fun(), mat_fun()))
+```
+
+### Drawbacks
+
+  **To-Do**
+  
+## 12. Use of Native Binary Formats for I/O
+
+### Idea
+Using a data type that is native to R should intuitively result in faster reading/loading times. As an added example, when the file are stored in native file formats, such as `.rds`, they are also quite heavily compressed, resulting in efficient usage of resources.
+
+### Code Examples
+
+#### Unoptimized Code
+
+```{r io_fun_code_og}
+mtcars$cyl <- as.factor(mtcars$cyl)
+
+io_fun <- function() {
+  write.csv(mtcars, file = "mtcars.csv")
+  read.csv("mtcars.csv")
+}
+```
+
+#### Proposed Optimized Code
+
+```{r io_fun_code_op}
+io_opt_fun <- function() {
+  saveRDS(mtcars, file = "mtcars.rds")  
+  readRDS("mtcars.rds")
+}
+```
+
+### Benchmark
+
+```{r io_fun_benchmark, warning = FALSE, message = FALSE}
+library("ggplot2")
+library("microbenchmark")
+autoplot(microbenchmark(io_fun(), io_opt_fun()))
+```
+
+### Drawbacks
+
+  **To-Do**
+  
+## 13. Efficient Column Extraction
+
+### Idea
+The idea would be to replace the different one-column extraction alternatives by the .subset2 function.
+
+### Code Examples
+
+#### Unoptimized Code
+
+```{r col_ext_og}
+col_ext_fun <- function(){
+  mtcars[, 11]
+  mtcars$carb
+  mtcars[[c(11)]]
+  mtcars[[11]]  
+}
+```
+
+#### Proposed Optimized Code
+
+```{r col_ext_op}
+col_ext_op_fun <- function(){
+  .subset2(mtcars, 11)
+  .subset2(mtcars, 11)
+  .subset2(mtcars, 11)
+  .subset2(mtcars, 11)
+}
+```
+
+### Benchmark
+
+```{r col_ext_benchmark, warning = FALSE, message = FALSE}
+library("ggplot2")
+library("microbenchmark")
+autoplot(microbenchmark(
+  mtcars[, 11],
+  mtcars$carb,
+  mtcars[[c(11)]],
+  mtcars[[11]],
+  .subset2(mtcars, 11)
+))
+```
+
+### Drawbacks
+
+  **To-Do**
+
+## 14. Efficient Value Extraction
+
+### Idea
+The idea would be to replace the different one-value extraction alternatives by the .subset2 function
+
+### Code Examples
+
+#### Unoptimized Code
+```{r val_ext_og}
+val_ext_fun <- function(){
+  mtcars[32, 11]
+  mtcars$carb[32]
+  mtcars[[c(11, 32)]]
+  mtcars[[11]][32]
+}
+```
+
+#### Proposed Optimized Code
+
+```{r val_ext_op}
+val_ext_op_fun <- function(){
+ .subset2(mtcars, 11)[32] 
+  .subset2(mtcars, 11)[32]
+  .subset2(mtcars, 11)[32]
+  .subset2(mtcars, 11)[32]
+}
+```
+### Benchmark
+
+```{r val_ext_benchmark, warning = FALSE, message = FALSE}
+library("ggplot2")
+library("microbenchmark")
+autoplot(microbenchmark(
+  mtcars[32, 11],
+  mtcars$carb[32],
+  mtcars[[c(11, 32)]],
+  mtcars[[11]][32],
+  .subset2(mtcars, 11)[32],
+  times = 1000L
+))
+```
+
+### Drawbacks
+
+  **To-Do**
+
+## External packages that can be explored for further speed-ups
```

Comment:
I would keep following lines just as comments for us, remove them from the vignette.
We could get ideas from these packages.